File: crates/source-languages/tests/fixtures/sample.dart
/// A tiny geometry library used as the Dart tags-query fixture.
library geometry;

import 'dart:math' as math;
import 'package:collection/collection.dart' show IterableExtension;

const double defaultScale = 2.0;

/// Anything that can report an area.
abstract class Shape {
  double get area;

  String describe();
}

mixin Loggable {
  final List<String> log = <String>[];

  void record(String message) {
    log.add(message);
  }
}

class Circle extends Shape with Loggable implements Comparable<Circle> {
  Circle(this.radius);

  factory Circle.unit() => Circle(1.0);

  final double radius;

  static const double tau = 2 * math.pi;

  @override
  double get area => math.pi * radius * radius;

  @override
  String describe() {
    record('circle');
    return 'circle of area ${area.toStringAsFixed(2)}';
  }

  @override
  int compareTo(Circle other) => area.compareTo(other.area);

  Circle operator +(Circle other) => Circle(radius + other.radius);
}

enum Palette { red, green, blue }

typedef Transform = double Function(double value);

extension ShapeList on List<Shape> {
  double get total => fold(0.0, (sum, shape) => sum + shape.area);
}

double scaleAll(List<Shape> shapes, Transform transform) {
  var sum = 0.0;
  for (final shape in shapes) {
    sum += transform(shape.area);
  }
  return sum;
}

void main() {
  final shapes = <Shape>[Circle(1.5), Circle.unit()];
  final scaled = scaleAll(shapes, (value) => value * defaultScale);
  print(shapes.first.describe());
  print('${shapes.total} $scaled ${Palette.red}');
}

File: crates/source-languages/tests/fixtures/sample.zig
//! A tiny geometry module used as the Zig tags-query fixture.

const std = @import("std");
const math = std.math;

pub const max_shapes: usize = 16;

pub const Error = error{ TooManyShapes, Degenerate };

/// A point in the plane.
pub const Point = struct {
    x: f64,
    y: f64,

    pub fn init(x: f64, y: f64) Point {
        return .{ .x = x, .y = y };
    }

    pub fn magnitude(self: Point) f64 {
        return math.sqrt(self.x * self.x + self.y * self.y);
    }
};

pub const Kind = enum {
    circle,
    rectangle,
};

pub const Shape = union(Kind) {
    circle: f64,
    rectangle: Point,
};

fn area(shape: Shape) f64 {
    return switch (shape) {
        .circle => |radius| math.pi * radius * radius,
        .rectangle => |corner| corner.x * corner.y,
    };
}

pub fn totalArea(shapes: []const Shape) Error!f64 {
    if (shapes.len > max_shapes) return Error.TooManyShapes;
    var total: f64 = 0;
    for (shapes) |shape| {
        total += area(shape);
    }
    return total;
}

pub fn main() !void {
    const origin = Point.init(0, 0);
    std.debug.print("{d}\n", .{origin.magnitude()});
    const shapes = [_]Shape{ .{ .circle = 1.0 }, .{ .rectangle = origin } };
    const total = try totalArea(&shapes);
    std.debug.print("{d}\n", .{total});
}

test "total area sums every shape" {
    const shapes = [_]Shape{.{ .circle = 1.0 }};
    try std.testing.expect(try totalArea(&shapes) > 3.0);
}
